tests/python_package_test/test_sklearn.py

```diff
@@ -117,6 +117,24 @@ def test_lambdarank(self):
         self.assertGreater(gbm.best_score_['valid_0']['ndcg@1'], 0.6333)
         self.assertGreater(gbm.best_score_['valid_0']['ndcg@3'], 0.6048)
 
+    def test_xendcg(self):
+        X_train, y_train = load_svmlight_file(os.path.join(os.path.dirname(os.path.realpath(__file__)),
+                                                           '../../examples/xendcg/rank.train'))
+        X_test, y_test = load_svmlight_file(os.path.join(os.path.dirname(os.path.realpath(__file__)),
+                                                         '../../examples/xendcg/rank.test'))
+        q_train = np.loadtxt(os.path.join(os.path.dirname(os.path.realpath(__file__)),
+                                          '../../examples/xendcg/rank.train.query'))
+        q_test = np.loadtxt(os.path.join(os.path.dirname(os.path.realpath(__file__)),
+                                         '../../examples/xendcg/rank.test.query'))
+        gbm = lgb.LGBMRanker(n_estimators=50, objective='rank_xendcg', random_state=31, n_jobs=1)
```

Comment:
I think we should get the same results but with better speed without limiting `n_jobs=1`.

Reply:
If I increased the number of jobs, the calls to generate the next random number may be invoked in a different order, leading to a different value of NDCG.

Reply:
@sbruch 
Ah, I see! The following line causes irreproducible results for number of threads different from 1 under the OMP loop, right?
```
gammas[i] = rand_->NextFloat();
```

We do receive a lot of issues related to the reproducibility and do our best in this direction. In case users can obtain reproducible results with **XE_NDCG_MART** objective function only if they set `num_threads` to 1, I think it's definitely worth to mention this fact in LightGBM's docs. For example, in `config.h`:
```
// descl2 = to obtain reproducible results, you should disable parallelism by setting ``num_threads`` to 1
```

And/or in FAQ, like it is done for **bagging**: https://lightgbm.readthedocs.io/en/latest/FAQ.html#bagging-is-not-reproducible-when-changing-the-number-of-threads.

cc @guolinke

Reply:
Correct. Sorry I should have explained better why num_threads>1 fails the test case.

That makes sense. I've added a note to that effect in config.h as requested.